include/display.h

```diff
@@ -0,0 +1,365 @@
+ */
+typedef int (*display_write_bitmap_api)(const struct device *dev, const u16_t x,
+					const u16_t y, const u16_t w,
+					const u16_t h, const u16_t p,
+					const void *data);
+
+/**
+ * @typedef display_read_bitmap_api
+ * @brief Callback API for reading bitmap
+ * See display_read_bitmap() for argument description
+ */
+typedef int (*display_read_bitmap_api)(const struct device *dev, const u16_t x,
+				       const u16_t y, const u16_t w,
+				       const u16_t h, const u16_t p,
+				       void *data);
+
+/**
+ * @typedef display_get_framebuffer_api
+ * @brief Callback API to get framebuffer pointer
+ * See display_get_framebuffer() for argument description
+ */
+typedef void *(*display_get_framebuffer_api)(const struct device *dev);
+
+/**
+ * @typedef display_set_brightness_api
+ * @brief Callback API to set display brightness
+ * See display_set_brightness() for argument description
+ */
+typedef int (*display_set_brightness_api)(const struct device *dev,
+					  const u8_t brightness);
+
+/**
+ * @typedef display_set_contrast_api
+ * @brief Callback API to set display contrast
+ * See display_set_contrast() for argument description
+ */
+typedef int (*display_set_contrast_api)(const struct device *dev,
+					const u8_t contrast);
+
+/**
+ * @typedef display_get_capabilities_api
+ * @brief Callback API to get display capabilities
+ * See display_get_capabilities() for argument description
+ */
+typedef void (*display_get_capabilities_api)(const struct device *dev,
+		struct display_capabilities *capabilities);
+
+/**
+ * @typedef display_set_pixel_format_api
+ * @brief Callback API to set pixel format used by the display
+ * See display_set_pixel_format() for argument description
+ */
+typedef int (*display_set_pixel_format_api)(const struct device *dev,
+		const enum display_pixel_format pixel_format);
+
+/**
+ * @typedef display_set_orientation_api
+ * @brief Callback API to set orientation used by the display
+ * See display_set_orientation() for argument description
+ */
+typedef int (*display_set_orientation_api)(const struct device *dev,
+		const enum display_orientation orientation);
+
+/**
+ * @brief Display driver API
+ * API which a display driver should expose
+ */
+struct display_driver_api {
+	display_on_api display_on;
+	display_off_api display_off;
+	display_write_bitmap_api write_bitmap;
+	display_read_bitmap_api read_bitmap;
+	display_get_framebuffer_api get_framebuffer;
+	display_set_brightness_api set_brightness;
+	display_set_contrast_api set_contrast;
+	display_get_capabilities_api get_capabilities;
+	display_set_pixel_format_api set_pixel_format;
+	display_set_orientation_api set_orientation;
+};
+
+/**
+ * @brief Write bitmap
+ *
+ * @param dev Pointer to device structure
+ * @param x x coordinate of the upper left corner
+ * @param y y coordinate of the upper left corner
+ * @param w width of the bitmap in pixels
+ * @param h height of the bitmap in pixels
+ * @param p pitch between lines in pixels
```

Comment:
Overall, only the parameter `pitch` disturbs me. As I see in [ili9340_write_bitmap](https://github.com/zephyrproject-rtos/zephyr/pull/7833/commits/bf5f5a89af270bc4e5d98ae59d41cf7bcadcfef8#diff-8e80313f451c3259257eeb8f023217fcR162) the same can be achieved with sequential calls of display_write_bitmap with approximately the same performance. If `pitch` is necessary, why can not it be realized through higher layers, why should every driver handle it?

Reply:
This parameter was requested by @pizi-nordic and as the impact to the driver is minimal, splitting a single  transfer in multiple transfers, it was added giving the higher layers more flexibility.

E.g. if you perform a image transformation on 30x30 of data you could transfer already 10x10 data once it is ready. Of course the higher layer could split it in 10 calls of 10x1, but for the current ili9340 driver this results in more SPI transfers as it reposition the window for every call.

Reply:
>Of course the higher layer could split it in 10 calls of 10x1, but for the current ili9340 driver this results in more SPI transfers as it reposition the window for every call.

It would be acceptable for me as it keeps the drivers simple. Otherwise, I fear higher probability for the errors, there is also no possibility to check whether calculated pointer is still within the limits of the buffer.

Reply:
True that there is no way to check the buffer limit but it is also the case without the p parameter, the only solution here is to add yet another parameter to the function that contains the buffer length.

Reply:
What do you think about separating the function `display_write/display_read` into `display_set_mem_region(x, y, w, h)` and `display_write(ptr, len)/dispay_read(ptr, len)`? 

_I hope you still have patience with me :-)_

Reply:
I prefer to keep it as-is this would just complicate the higher layer software (2 calls needed) and make the assumption that these calls can be separated (you never know what kind of devices are out there). In case there is a device that can only update the hole screen at once safety checks  can be added added to see if the caller complies with this limitation.

> I hope you still have patience with me :-)

_For the moment, yes :P 
I hope it is likewise_

Reply:
>In case there is a device that can only update the hole screen at once safety checks can be added added to see if the caller complies with this limitation.

Then display_set_mem_region would return a error.  btw, something like a display_get_mem_region could return the smallest region size.

Reply:
+1 to keep display_read/display_write as is. We have multi-thread system and with these calls separated, we have to be sure that nobody else will touch our display between memory region set and read/write.

Reply:
>We have multi-thread system and with these calls separated, we have to be sure that nobody else will touch our display between memory region set and read/write.

Only out of curiosity what possible situation would it be?

Reply:
>> We have multi-thread system and with these calls separated, we have to be sure that nobody else will touch our display between memory region set and read/write.
>
>Only out of curiosity what possible situation would it be?

Consider a device which have UI similar to any android phone: on the top of the screen you have your battery level indicator, clock and so on. These are always displayed and whole GUI is below.

In such setup we might have one thread implementing GUI and other one(s) updating clock and battery indicator. As they are rendering to different parts of screen, there is no need of any synchronization, except display access. So for example the thread updating battery level could wake up every 30s, read battery level, render the indicator and update it on screen.

If this update can be done using only one call you can just have lock inside driver. If you have multiple calls, there has to be external locking.

Reply:
>If you have multiple calls, there has to be external locking.

Would not it be better to implement a Display Region Manager as an intermediate layer and keep the driver simple?

Reply:
In the provided example, this is done by design. No other thread is touching battery indicator region except the one handling batter measurement.

Reply:
It is not. With this API, as it is, what prevents the thread from writing to any region?

Reply:
The API of display driver does not prevent from accessing same error by multiple threads.
In the example I proposed such thing is done by design of the example, not driver.

Of course, we could add upper layer which will provide display region reservation, but I think that such thing maybe an overkill for embedded system.

Reply:
>From the point of view of the higher layer software windowing is a useful feature for performance/memory reasons and lot of code will just result in calling display_set_mem_region followed by a call to display_write instead of a single call to display_write.

Your suggestion of the display_write will do the same, set scanning region and then write to controller ram. I do not see any big impact on performance. My point remains: the function is too error prone. Which interface is safer: `display_set_mem_region(dev, x, y, w, h) followed by display_write(dev, buf, len)/display_write(dev, buf, len, stride)` or `display_write(dev, x, y, w, h, p, buf)`? Has no one bad feeling with `display_read(dev, x, y, w, h, p, buf)`?

Reply:
>What about adding the len param to the proposed API and put the x, y, w and h in
a struct display_write(dev, dim, buf, len, p)?

Alternatively, you can insert a structure that describes the display buffer (and perhaps the selected region). The buffer may also be larger than the memory of the display, e.g. to realize text scrolling.

Reply:
@jfischer-phytec-iot updated api with a buffer descriptor